src/lib.rs

```diff
@@ -297,13 +238,101 @@ impl AsyncHttpRangeReader {
 
     /// Initialize the reader from [`AsyncHttpRangeReader::initial_head_request`] (or a user
     /// provided response the)
-    pub async fn from_head_response(
+    pub fn from_head_response(
         client: impl Into<reqwest_middleware::ClientWithMiddleware>,
         head_response: Response,
         extra_headers: HeaderMap,
     ) -> Result<Self, AsyncHttpRangeReaderError> {
-        let client = client.into();
+        AsyncHttpRangeReaderBuilder::from_head_response(head_response)?
+            .client(client.into())
+            .extra_headers(extra_headers)
+            .build()
+            .map_err(AsyncHttpRangeReaderError::from)
+    }
+
+    /// Returns the ranges that this instance actually performed HTTP requests for.
+    pub async fn requested_ranges(&self) -> Vec<Range<u64>> {
+        let mut inner = self.inner.lock().await;
+        if let Some(Some(new_state)) = inner.streamer_state_rx.next().now_or_never() {
+            inner.streamer_state = new_state;
+        }
+        inner.streamer_state.requested_ranges.clone()
+    }
+
+    /// Prefetches a range of bytes from the remote. When specifying a large range this can
+    /// drastically reduce the number of requests required to the server.
+    pub async fn prefetch(&mut self, bytes: Range<u64>) {
+        let inner = self.inner.get_mut();
+
+        // Ensure the range is withing the file size and non-zero of length.
+        let range = bytes.start..(bytes.end.min(inner.data.len() as u64));
+        if range.start >= range.end {
+            return;
+        }
+
+        // Check if the range has been requested or not.
+        let inner = self.inner.get_mut();
+        if let Some((new_range, _)) = inner.requested_range.cover(range.clone()) {
+            let _ = inner.request_tx.send(range).await;
+            inner.requested_range = new_range;
+        }
+    }
+
+    /// Returns the length of the stream in bytes
+    #[allow(clippy::len_without_is_empty)]
+    pub fn len(&self) -> u64 {
+        self.len
+    }
+}
 
+#[derive(Default)]
+pub struct AsyncHttpRangeReaderBuilder {
+    client: Option<reqwest_middleware::ClientWithMiddleware>,
+    url: Option<Url>,
+    extra_headers: HeaderMap,
+    requested_range: SparseRange,
+    streamer_state: StreamerState,
+    initial_tail_response: Option<(Response, u64)>,
+    content_length: usize,
+}
+
+impl AsyncHttpRangeReaderBuilder {
+    pub fn client(mut self, client: reqwest_middleware::ClientWithMiddleware) -> Self {
+        self.client = Some(client);
+        self
+    }
+
+    pub fn url(mut self, url: Url) -> Self {
+        self.url = Some(url);
+        self
+    }
+
+    pub fn extra_headers(mut self, extra_headers: HeaderMap) -> Self {
+        self.extra_headers = extra_headers;
+        self
+    }
+
+    pub fn requested_range(mut self, requested_range: SparseRange) -> Self {
+        self.requested_range = requested_range;
+        self
+    }
```

Comment:
This is only used internally when calling `from_tail_response`. It should not be set by the user.

Reply:
Sounds good. I will change it

Reply:
I ended up thinking the same about `initial_tail_response` (seems like only `from_tail_response` needs to call that?)